File: Diffie-Hellman key exchange/Diffie_Hellman.py
import random

def generate_private_key(n, g):
    # Alice private key
    a = random.randint(1,n)

    # Bob private key
    b = random.randint(1, n)

    # gˆa mod n, Alice's k1
    k1 = pow(g, a, n)

    # gˆb mod n, Bob's k2
    k2 = pow(g, b, n)

    # If an atacker try to get x and y (the private keys) => the discrete logarithm problem
    # It is exponentially slow process
    print("Alice's shared key: ", pow(k2, a, n))
    print("Bob's shared key: ", pow(k1, b, n))

if __name__ == '__main__':
    # Should be a huge prime number
    n = 37
    # Primitive root of n
    g = 13

    # Using Diffie-Hellman cryptosystem to generate the shared key (for DES and AES)
    generate_private_key(n, g)

File: Diffie-Hellman key exchange/Diffie_Hellman.java
import java.security.SecureRandom;
import java.math.BigInteger;

public class Diffie_Hellman{
    private SecureRandom a, b;
    private BigInteger k1, k2, sharedKeyAlice, sharedKeyBob;

    public Diffie_Hellman(int n, int g){
        BigInteger G = BigInteger.valueOf(g);
        BigInteger N = BigInteger.valueOf(n);

        // Alice private key
        a = new SecureRandom();

        // Bob private key
        b = new SecureRandom();

        int numBits = 128;
        BigInteger exponential_a = new BigInteger(numBits, a);
        BigInteger exponential_b = new BigInteger(numBits, b);


        // gˆa mod n, Alice's k1
        k1 = G.modPow(exponential_a, N);

        // gˆb mod n, Alice's k1
        k2 = G.modPow(exponential_b, N);

        sharedKeyAlice = k2.modPow(exponential_a, N); // (g^b)^a mod n
        sharedKeyBob = k1.modPow(exponential_b, N);   // (g^a)^b mod n
    }

    public BigInteger getSharedKeyAlice() {
        return sharedKeyAlice;
    }

    public BigInteger getSharedKeyBob() {
        return sharedKeyBob;
    }


    public static void main(String[] args) {
        int n = 37; // prime number
        int g = 13;   // primitive root of n

        Diffie_Hellman dh = new Diffie_Hellman(n, g);

        System.out.printf("Shared key (Alice): %s\n", dh.getSharedKeyAlice());
        System.out.printf("Shared key (Bob)  : %s\n", dh.getSharedKeyBob());
    }
}